### PR TITLE
Testing: Centralize image resolution

### PR DESCRIPTION
**Pull request by snazy**
* Add `ContainerSpecHelper` to help resolving a Docker image
* Add support for "pinned" image references (including "sha256")

### CODE REVIEW

**Comment by snazy** [comment]
Required for #8541

File: testing/container-spec-helper/src/main/java/org/projectnessie/nessie/testing/containerspec/ContainerSpecHelper.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+  public abstract Class<?> containerClass();
+
+  public DockerImageName dockerImageName(String explicitImageName) {
+    if (explicitImageName != null) {
+      return DockerImageName.parse(explicitImageName);
+    }
+
+    String dockerfile = format("Dockerfile-%s-version", name());
+    URL resource = containerClass().getResource(dockerfile);
```
Comment (adutra) [inline testing/container-spec-helper/src/main/java/org/projectnessie/nessie/testing/containerspec/ContainerSpecHelper.java:59]: `containerClass()` is only used to serve as the class loader to use here. Isn't possible to just use `this.getClass()` instead?

File: versioned/storage/jdbc-tests/src/main/java/org/projectnessie/versioned/storage/jdbctests/CockroachBackendTestFactory.java [hunk 2]
```diff
@@ -30,6 +30,7 @@ public String getName() {
   @Nonnull
   @Override
   protected JdbcDatabaseContainer<?> createContainer() {
-    return new CockroachContainer(dockerImage("cockroach"));
+    return new CockroachContainer(
+        dockerImage("cockroach").asCompatibleSubstituteFor("cockroachdb/cockroach"));
```
Comment (adutra) [inline versioned/storage/jdbc-tests/src/main/java/org/projectnessie/versioned/storage/jdbctests/CockroachBackendTestFactory.java:34]: What motivates the addition of `.asCompatibleSubstituteFor()`? I've seen it added in a few places, but not everywhere.

File: testing/container-spec-helper/src/main/java/org/projectnessie/nessie/testing/containerspec/ContainerSpecHelper.java (hunk 1 above)
Comment (snazy) [reply testing/container-spec-helper/src/main/java/org/projectnessie/nessie/testing/containerspec/ContainerSpecHelper.java:59]: That would result in a re-build and re-test of all modules that use a container for all unrelated image version bumps - not a fan of that.

File: versioned/storage/jdbc-tests/src/main/java/org/projectnessie/versioned/storage/jdbctests/CockroachBackendTestFactory.java (hunk 2 above)
Comment (snazy) [reply versioned/storage/jdbc-tests/src/main/java/org/projectnessie/versioned/storage/jdbctests/CockroachBackendTestFactory.java:34]: Required by the testcontainer classes.

File: testing/container-spec-helper/src/main/java/org/projectnessie/nessie/testing/containerspec/ContainerSpecHelper.java (hunk 1 above)
Comment (snazy) [reply testing/container-spec-helper/src/main/java/org/projectnessie/nessie/testing/containerspec/ContainerSpecHelper.java:59]: Yea - that comment was for the other thread.
We have to do one thing - either use an "absolute path" to the Dockerfile or a `Class` +name tuple (note: we need the name anyways).

Comment (snazy) [reply testing/container-spec-helper/src/main/java/org/projectnessie/nessie/testing/containerspec/ContainerSpecHelper.java:59]: Note: having all Dockerfiles in the "same directory" would yield duplicate resources - causing warnings and potentially conflicts later.

**Review by adutra — DISMISSED** [review]

**Review by adutra — APPROVED** [review]

**Pull request merged**